pkg/dataloader/releaseloader/releasesync.go

```diff
@@ -104,98 +109,114 @@ func (r *ReleaseLoader) Load() {
 	}
 }
 
-func (r *ReleaseLoader) buildReleaseTag(architecture, release string, tag ReleaseTag) *models.ReleaseTag {
-	releaseDetails := r.fetchReleaseDetails(architecture, release, tag)
-	releaseTag := releaseDetailsToDB(architecture, tag, releaseDetails)
+func (r *ReleaseLoader) buildReleaseTag(rs ReleaseStream, tag ReleaseTag) *models.ReleaseTag {
+	releaseDetails := r.fetchReleaseDetails(rs, tag)
+	if releaseDetails == nil {
+		return nil
+	}
+	releaseTag := releaseDetailsToDB(rs, tag, *releaseDetails)
 
 	// We skip releases that aren't fully baked (i.e. all jobs run and changelog calculated)
 	if releaseTag == nil || (releaseTag.Phase != api.PayloadAccepted && releaseTag.Phase != api.PayloadRejected) {
 		return nil
 	}
 
-	// PR is many-to-many, find the existing relation. TODO: There must be a more clever way to do this...
-	for i, pr := range releaseTag.PullRequests {
-		existingPR := models.ReleasePullRequest{}
-		result := r.db.DB.Table("release_pull_requests").Where("url = ?", pr.URL).Where("name = ?", pr.Name).First(&existingPR)
-		if result.Error == nil {
-			releaseTag.PullRequests[i] = existingPR
-		}
+	if len(releaseTag.PullRequests) > 0 {
+		releaseTag.PullRequests = r.resolveReleasePullRequests(releaseTag.PullRequests)
 	}
 
 	return releaseTag
 }
 
-func (r *ReleaseLoader) fetchReleaseDetails(architecture, release string, tag ReleaseTag) ReleaseDetails {
-	releaseDetails := ReleaseDetails{}
-	releaseName := release
-	if architecture != "amd64" {
-		releaseName += "-" + architecture
+func (r *ReleaseLoader) resolveReleasePullRequests(pullRequests []models.ReleasePullRequest) []models.ReleasePullRequest {
+	if len(pullRequests) == 0 {
+		return pullRequests
+	}
+
+	type prKey struct{ url, name string }
+	prIndexMap := make(map[prKey]int, len(pullRequests))
+	orConditions := make([]string, 0, len(pullRequests))
+	args := make([]any, 0, len(pullRequests)*2)
```

Comment:
Did you consider defining a new struct that contains all the info we need rather than using multiple maps and keeping them in sync with an index map?

Reply:
TBH I didn't write this part and only gave it a cursory look... now that you mention it, this has a code smell for me, there's rarely a good reason to look at indexes. There's only one map, but it just ought to map to the PRs.